File: LudoGameAndroid/app/src/main/java/com/ludogame/app/engine/LudoBoardLayout.kt
package com.ludogame.app.engine

import androidx.compose.ui.geometry.Offset
import com.ludogame.app.models.BoardCorner
import com.ludogame.app.models.GameToken
import com.ludogame.app.models.PlayerColor
import kotlin.math.cos
import kotlin.math.sin

object LudoBoardLayout {
    const val PATH_LENGTH = 52
    const val HOME_STRETCH_LENGTH = 6

    val mainTrack: List<Offset> = run {
        val inset = 0.18f
        val end = 1.0f - inset
        val span = end - inset

        (0 until PATH_LENGTH).map { index ->
            val side = index / 13
            val step = index % 13
            val t = step / 12f

            when (side) {
                0 -> Offset(inset + span * t, end)
                1 -> Offset(end, end - span * t)
                2 -> Offset(end - span * t, inset)
                else -> Offset(inset, inset + span * t)
            }
        }
    }

    val centerCoordinate = Offset(0.5f, 0.5f)

    fun absoluteTrackIndex(token: GameToken): Int? {
        if (!token.isOnMainTrack) return null
        return (token.owner.startPosition + token.position) % PATH_LENGTH
    }

    fun coordinate(token: GameToken, stackIndex: Int = 0): Offset {
        return when {
            token.isInYard -> yardCoordinate(token.owner, token.id)
            token.isFinished -> centerCoordinate
            token.isInHomeStretch -> homeStretchCoordinate(
                token.owner,
                token.position - PATH_LENGTH,
                stackIndex
            )
            else -> {
                val absolute = absoluteTrackIndex(token)
                if (absolute != null) {
                    offset(mainTrack[absolute], stackIndex)
                } else {
                    centerCoordinate
                }
            }
        }
    }

    fun coordinateAtAbsoluteTrackIndex(index: Int): Offset {
        return mainTrack[index % PATH_LENGTH]
    }

    fun yardCoordinate(color: PlayerColor, tokenId: Int): Offset {
        val base = when (color.homeCorner) {
            BoardCorner.BOTTOM_LEFT -> Offset(0.09f, 0.91f)
            BoardCorner.TOP_LEFT -> Offset(0.09f, 0.09f)
            BoardCorner.TOP_RIGHT -> Offset(0.91f, 0.09f)
            BoardCorner.BOTTOM_RIGHT -> Offset(0.91f, 0.91f)
        }
        val offsets = listOf(
            Offset(-0.03f, -0.03f),
            Offset(0.03f, -0.03f),
            Offset(-0.03f, 0.03f),
            Offset(0.03f, 0.03f)
        )
        val offset = offsets[tokenId]
        return Offset(base.x + offset.x, base.y + offset.y)
    }

    private fun homeStretchCoordinate(color: PlayerColor, step: Int, stackIndex: Int): Offset {
        val clampedStep = step.coerceIn(0, HOME_STRETCH_LENGTH)
        val t = (clampedStep + 1).toFloat() / (HOME_STRETCH_LENGTH + 2)
        val start = color.homeCorner.entryPoint
        val end = centerCoordinate
        val point = Offset(
            x = start.x + (end.x - start.x) * t,
            y = start.y + (end.y - start.y) * t
        )
        return offset(point, stackIndex)
    }

    private fun offset(point: Offset, stackIndex: Int): Offset {
        val angle = stackIndex * Math.PI.toFloat() / 2f
        val radius = 0.012f
        return Offset(
            x = point.x + cos(angle) * radius,
            y = point.y + sin(angle) * radius
        )
    }

    private val BoardCorner.entryPoint: Offset
        get() = when (this) {
            BoardCorner.BOTTOM_LEFT -> Offset(0.38f, 0.62f)
            BoardCorner.TOP_LEFT -> Offset(0.38f, 0.38f)
            BoardCorner.TOP_RIGHT -> Offset(0.62f, 0.38f)
            BoardCorner.BOTTOM_RIGHT -> Offset(0.62f, 0.62f)
        }
}

File: LudoGame/LudoGame/Engine/LudoBoardLayout.swift
import CoreGraphics
import Foundation

/// Maps logical board positions to normalized coordinates (0...1).
enum LudoBoardLayout {
    static let pathLength = 52
    static let homeStretchLength = 6

    /// Outer track cell centers, indexed 0...51 clockwise from red's start.
    static let mainTrack: [CGPoint] = {
        let inset: CGFloat = 0.18
        let end: CGFloat = 1.0 - inset
        let span = end - inset

        return (0..<pathLength).map { index in
            let side = index / 13
            let step = index % 13
            let t = CGFloat(step) / 12.0

            switch side {
            case 0:
                return CGPoint(x: inset + span * t, y: end)
            case 1:
                return CGPoint(x: end, y: end - span * t)
            case 2:
                return CGPoint(x: end - span * t, y: inset)
            default:
                return CGPoint(x: inset, y: inset + span * t)
            }
        }
    }()

    static func absoluteTrackIndex(for token: GameToken) -> Int? {
        guard token.isOnMainTrack else { return nil }
        return (token.owner.startPosition + token.position) % pathLength
    }

    static func coordinate(for token: GameToken, stackIndex: Int = 0) -> CGPoint {
        if token.isInYard {
            return yardCoordinate(for: token.owner, tokenID: token.id)
        }
        if token.isFinished {
            return centerCoordinate
        }
        if token.isInHomeStretch {
            return homeStretchCoordinate(for: token.owner, step: token.position - pathLength, stackIndex: stackIndex)
        }
        if let absolute = absoluteTrackIndex(for: token) {
            return offset(point: mainTrack[absolute], stackIndex: stackIndex)
        }
        return centerCoordinate
    }

    static func coordinate(atAbsoluteTrackIndex index: Int) -> CGPoint {
        mainTrack[index % pathLength]
    }

    static var centerCoordinate: CGPoint {
        CGPoint(x: 0.5, y: 0.5)
    }

    static func yardCoordinate(for color: PlayerColor, tokenID: Int) -> CGPoint {
        let base: CGPoint
        switch color.homeCorner {
        case .bottomLeft: base = CGPoint(x: 0.09, y: 0.91)
        case .topLeft: base = CGPoint(x: 0.09, y: 0.09)
        case .topRight: base = CGPoint(x: 0.91, y: 0.09)
        case .bottomRight: base = CGPoint(x: 0.91, y: 0.91)
        }
        let offsets: [CGPoint] = [
            CGPoint(x: -0.03, y: -0.03),
            CGPoint(x: 0.03, y: -0.03),
            CGPoint(x: -0.03, y: 0.03),
            CGPoint(x: 0.03, y: 0.03)
        ]
        return CGPoint(x: base.x + offsets[tokenID].x, y: base.y + offsets[tokenID].y)
    }

    private static func homeStretchCoordinate(for color: PlayerColor, step: Int, stackIndex: Int) -> CGPoint {
        let clampedStep = min(max(step, 0), homeStretchLength)
        let t = CGFloat(clampedStep + 1) / CGFloat(homeStretchLength + 2)
        let start = color.homeCorner.entryPoint
        let end = centerCoordinate
        let point = CGPoint(
            x: start.x + (end.x - start.x) * t,
            y: start.y + (end.y - start.y) * t
        )
        return offset(point: point, stackIndex: stackIndex)
    }

    private static func offset(point: CGPoint, stackIndex: Int) -> CGPoint {
        let angle = CGFloat(stackIndex) * .pi / 2
        let radius: CGFloat = 0.012
        return CGPoint(
            x: point.x + cos(angle) * radius,
            y: point.y + sin(angle) * radius
        )
    }
}

private extension BoardCorner {
    var entryPoint: CGPoint {
        switch self {
        case .bottomLeft: return CGPoint(x: 0.38, y: 0.62)
        case .topLeft: return CGPoint(x: 0.38, y: 0.38)
        case .topRight: return CGPoint(x: 0.62, y: 0.38)
        case .bottomRight: return CGPoint(x: 0.62, y: 0.62)
        }
    }
}
